mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp

```diff
@@ -1100,6 +1103,408 @@ mlir::scf::tileConsumerAndFuseProducersUsingSCF(
+    rank = forallOp.getRank();
+  }
+
+  if (failed(checkAssumptionForLoop(oldLoopOp, consumerOp))) {
+    return rewriter.notifyMatchFailure(
+        oldLoopOp, "containing loop op should either yield just one value or "
+                   "have the consumer op as its first user");
+  }
+
+  OpBuilder::InsertionGuard g(rewriter);
+
+  // 2. Check consumer is not using scf loop's output as init.
+  auto dstOp = cast<DestinationStyleOpInterface>(consumerOp);
+  SmallVector<Value> dpsInits =
+      llvm::map_to_vector(dstOp.getDpsInits(), [](Value v) { return v; });
+  if (llvm::is_contained(dpsInits, oldLoopOp->getResult(resultNumber))) {
+    return rewriter.notifyMatchFailure(
+        consumerOp,
+        "consumer op taking the result of scf.for as init is not supported");
+  }
+  newOuts.append(dpsInits);
+
+  Location loc = oldLoopOp->getLoc();
+
+  // 3. Create new scf loop op.
+  rewriter.setInsertionPoint(consumerOp);
+  Operation *newLoopOp = nullptr;
+  Block *newLoopBody = nullptr;
+  if (isInsertSliceOp) {
+    auto forOp = cast<scf::ForOp>(oldLoopOp);
+    auto newForOp = rewriter.create<scf::ForOp>(loc, forOp.getLowerBound(),
+                                                forOp.getUpperBound(),
+                                                forOp.getStep(), newOuts);
+    newLoopOp = newForOp;
+    newLoopBody = newForOp.getBody();
+  } else {
+    auto forallOp = cast<scf::ForallOp>(oldLoopOp);
+    auto newForallOp = rewriter.create<scf::ForallOp>(
+        loc, forallOp.getMixedLowerBound(), forallOp.getMixedUpperBound(),
+        forallOp.getMixedStep(), newOuts, forallOp.getMapping());
+    newLoopOp = newForallOp;
+    rewriter.eraseOp(newForallOp.getTerminator());
+    newLoopBody = newForallOp.getBody();
+  }
+
+  // 4. Move the loop body to the new op.
+  unsigned oldNumArguments = oldLoopBody->getNumArguments();
+  rewriter.mergeBlocks(oldLoopBody, newLoopBody,
+                       newLoopBody->getArguments().take_front(oldNumArguments));
+
+  // 5.a. Clone consumer after the cloned
+  // tensor.insert_slice/parallel_insert_slice op.
+  rewriter.setInsertionPointAfter(candidateSliceOp);
+  auto newForOpBlockArgsForConsumerDest =
+      newLoopBody->getArguments().drop_front(oldNumArguments);
+  auto clonedConsumerOp = cast<TilingInterface>(cloneOpAndUpdateDestinationArgs(
+      rewriter, consumerOp, newForOpBlockArgsForConsumerDest));
+
+  // 5.b. Replace all uses of the loop result with the result of the cloned
+  // tensor.insert_slice/parallel_insert_slice.
+  OpOperand &operandToReplace = clonedConsumerOp->getOpOperand(operandNumber);
+  rewriter.modifyOpInPlace(clonedConsumerOp, [&]() {
+    if (auto sliceOp = dyn_cast<tensor::InsertSliceOp>(candidateSliceOp)) {
+      operandToReplace.set(sliceOp.getResult());
+    } else if (auto sliceOp =
+                   dyn_cast<tensor::ParallelInsertSliceOp>(candidateSliceOp)) {
+      operandToReplace.set(sliceOp.getSource());
```

Comment:
This seems off. Here the source of the `sliceOp` is the tiled operation, but this is used to replace the operand of the cloned consumer which is untiled. This is also potentially related to the `extract_slice` issue pointed out by @qedawkins  earlier. This is what I think this should be doing

1) Create a new `tensor.insert_slice` that represent the insertion. In the `scf.for` case this is a clone of the `candidateSliceOp`. In the `scf.forall` case this is created from the `tensor.parallel_insert_slice` except now it has a result.
2) Clone the consumer op (already done here). Replace the operand in the consumer along which fusion is happening with the result of the `tensor.insert_slice`.
3) Call the method `tensor::replaceInsertSliceWithTiledConsumer` with the cloned insert slice op and the cloned consumer op
4) Now replace all `tensor.extract_slice` uses of the cloned `tensor.insert_slice` with the source of the cloned `tensor.insert_slice`. By construction all the slice uses of the `tensor.insert_slice` are exactly the same shape as the source.

Reply:
Thank you so much @MaheshRavishankar ! This made so much sense and has really helped clean up the issue with `tensor.extract_slice`! 

I've addressed it in the latest push.

Reply:
> By construction all the slice uses of the tensor.insert_slice are exactly the same shape as the source.

Where does this guarantee come from? Looking at the implementation of `replaceInsertSliceWithTiledConsumer` it is just calling `getTiledImplementationFromOperandTile` without passing in the source of the `insert_slice`.

Reply:
Adding here an explanation in case it helps :-

State of the IR after step 3 of the algo in the above thread :
```
         %clone_insert_slice = cloned tensor.insert_slice :
                                    %source<32> into %dest<64> | OFFSET | STRIDES | SIZES
                           
         %tiled_operand = tensor.extract_slice:
                           %clone_insert_slice<64> to <32> | OFFSET | STRIDES | SIZES
          
         %res = tiled_consumer_op %tiled_operand
```
NOTE:
1. OFFSETS | STRIDES | SIZES of both `tensor.insert_slice` and `tensor.extract_slice` is SAME, therefore it is a valid intermediate IR yielded by the `getTiledImplementationFromOperandTile` API which is in fact taking those OFFSETS/STRIDES info. And this basically means that `tensor.insert_slice` and `tensor.extract_slice` is just inverse of each other.
2. Right before calling the API to generate tiled consumer op, we have
   `%res = consumer_op %clone_insert_slice<64>`. 

Now, perhaps what @MaheshRavishankar meant above for step 4 is that by design the use of `%tiled_operand` is going to be exactly the same shape as the "source" of `%clone_insert_slice`. Thus we can simply replace all uses of `%tiled_operand` with `%source`.

Reply:
My main question is where in the tiling interface do we require that the tiled implementation generates this exact `%tiled_operand = tensor.extract_slice | OFFSET | STRIDES | SIZES` extract slice? The only requirement I see in `getTiledImplementation` is that it produces a result tile that corresponds to the given offsets/sizes/strides. I don't see any requirements given about how the operands are used.

Reply:
Oh okay.
Well, not sure about the doc comments there, but implementation-wise `getTiledImplementation` does in fact operate on the operands too via [makeTiledShapes here](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp#L123-L124).

Reply:
That's for linalg. Other operations can do whatever they want (not trying to be difficult, just trying to make sure we at least clarify any implicit interface API requirements like this).

Reply:
> > By construction all the slice uses of the tensor.insert_slice are exactly the same shape as the source.
> 
> Where does this guarantee come from? Looking at the implementation of `replaceInsertSliceWithTiledConsumer` it is just calling `getTiledImplementationFromOperandTile` without passing in the source of the `insert_slice`.

You are using the slice of the operand to compute the slice of the iteration space that computes that operand. That is only possible if that is a bijection. Then the tile of the operand computed from the iteration space is going to be the same. Things have to be consistent.

Yes, we are indexing on `tensor.extract_slice` and `tensor.insert_slice`, which we can maybe eventually switch to an interface that allows different kind of "slices", but this is the current state right now. If we miss it, then its a "missed optimization" and not a correctness issue.

Reply:
Ok I'll defer to you then, but I think it would be good to document this implicit dependence on `extract_slice` on the SCF api.